[Fornax-Mizar-Integration] Added gateway and virtual to subnet in data plane

jshaofuturewei opened the pull request:
There are following changes
1. Since we added virtual and cluster gateway to control plane in https://github.com/CentaurusInfra/mizar/pull/582, we made the data plane changes as well in trn cli to get virtual and cluster values. Tests are added as well.

2. There are changes in transit xdp:
a. Don't update cache if the requests come from virtual subnets. The request might come from remote clusters. It is not needed to update cache to set up direct path.

b. If the request targets are cluster gateway, pass the request to user space gateway agents. Since all the virtual subnets do not have endpoints in the the cluster, all the requests aiming virtual subnets have to pass to a new gateway agents to process.

The change will not affect the current mizar logic if the requests are not virtual subnet related.
Detailed tests have to wait when gateway agent is merged. 

Manual test steps:
1. Create a cluster with 2 worker nodes and 1 cluster gateway.
2. Create a vpc vpc1 and one local subnet  net1    and one virtual subnet net2
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# kubectl get subnets
NAME   IP              PREFIX   VNI       VPC    STATUS        VIRTUAL   BOUNCERS   CREATETIME                   PROVISIONDELAY
net0   20.0.0.0        8        1         vpc0   Provisioned   false     1          2022-01-12T21:53:13.216790   81.060556
net1   192.168.0.0     24       9044412   vpc1   Provisioned   false     1
net2   192.168.122.0   24       9044412   vpc1   Provisioned   true      1
```
3. Create a pod in subnet net1
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# kubectl get pods pod-in-net1 -o wide
NAME          READY   STATUS    RESTARTS   AGE     IP            NODE              NOMINATED NODE   READINESS GATES
pod-in-net1   1/1     Running   0          6m34s   192.168.0.6   ip-172-31-6-231   <none>           <none>
```
4. Ping a virtual subnet ip 
```bash
root@ip-172-31-7-100:~/mizar_cluster_scripts# ./blind_ping_122.sh
-------
pinging 192.168.122.13 from 192.168.0.6
-------
PING 192.168.122.13 (192.168.122.13) 56(84) bytes of data.
```
5. Verify the request targeting gateway host  is redirected to user space
```bash
cat /sys/kernel/debug/tracing/trace_pipe | tee out
```
According to the code changes in src/xdp/trn_transit_xdp.c, if the request is redirect to user spaces, it will print out "bpf_debug("[Transit:%d:0x%x] Edge gateway\n", __LINE__, bpf_ntohl(pkt->inner_ip->daddr));" as below

```bash
          <idle>-0     [000] ..s. 85815.136730: 0: [Transit:475:0xc0a87a0d] Edge gateway
          <idle>-0     [000] ..s. 85816.142070: 0: [Transit:751:0xac1f02d9] Processing IP
          <idle>-0     [000] ..s. 85816.142106: 0: [Transit:475:0xc0a87a0d] Edge gateway
          <idle>-0     [000] ..s. 85817.165908: 0: [Transit:751:0xac1f02d9] Processing IP
          <idle>-0     [000] ..s. 85817.165943: 0: [Transit:475:0xc0a87a0d] Edge gateway
```

jshaofuturewei committed:
1. Added cluster_gateway and virtual to the net object in trn cli
2. Did not update cache if the requests came from remote clusters
3. If the request targets are cluster gateway, redirect those requests to gateway

vinaykul commented:
Could you please add a description for the change that explains what the change is, and why it is needed. thanks.

jshaofuturewei commented:
> Could you please add a description for the change that explains what the change is, and why it is needed. thanks.



> Could you please add a description for the change that explains what the change is, and why it is needed. thanks.

I added the detailed description  and explain why we need it.

pdgetrf commented on src/cli/trn_cli_common.c:
```diff
@@ -151,6 +153,27 @@ int trn_cli_parse_net(const cJSON *jsonobj, struct rpc_trn_network_t *net)
 	}
 	net->switches_ips.switches_ips_len = i;
+	// Cluster gateway is NULL or empty if it is not set in a config map which is a string value
+	net->cluster_gateway = 0;
```
use "null" instead of "0"

jshaofuturewei replied:
It is an integer which can't be replace with null/NULL.

pdgetrf replied:
oh i see. I was confused by the comment on line 156.

pdgetrf commented on src/cli/trn_cli_common.c:
```diff
@@ -151,6 +153,27 @@ int trn_cli_parse_net(const cJSON *jsonobj, struct rpc_trn_network_t *net)
+		net->cluster_gateway = sa.sin_addr.s_addr;
+	}
+        // Virtual can be NULL if it is a local subnet. 0 as false and 1 as true
+	net->virtual = 0;
```
why 0 is used instead of true?

jshaofuturewei replied:
There is no bool type in the linux kernel.

pdgetrf replied:
got it.

pdgetrf commented on src/cli/trn_cli_common.c:
```diff
@@ -151,6 +153,27 @@ int trn_cli_parse_net(const cJSON *jsonobj, struct rpc_trn_network_t *net)
+				net->virtual = 1;
+			} else {
+				net->virtual = 0;
+			}
```
I feel this if block starting on line 166 can be replaced by 

```c
if (cJSON_IsBool(virtual) && cJSON_IsFalse(virtual) == 0) {
     net->virtual = 1;
} 
```

jshaofuturewei replied:
We have to check if the virtual is NULL since the incoming json request might not have virtual fields.

pdgetrf replied:
understand. I'm just trying to remove the most inner if block.

net->virtual is already set to 0 on line 164.

so the if block starting on line 166 (not 165) only needs to set net->virtual to 1 if  "cJSON_IsBool(virtual) && cJSON_IsFalse(virtual) == 0" right?

pdgetrf replied:
oh but the print_err on line 173 might need to be re-positioned. 

see if it's possible to reduce this section from 3-level if statement to at least 2-level as an optimization.

jshaofuturewei replied:
Updated

pdgetrf commented on src/xdp/trn_transit_xdp.c:
```diff
@@ -454,6 +457,26 @@ static __inline int trn_process_inner_ip(struct transit_packet *pkt)
+
+	// "__u32 nip[3]" has 12 bytes altogether
+	// tunnel_id is bit-wise 64 bit and occupies the first 8 bytes
+	// tip (target inner ip) occupies the rest 4 bytes
```
line 467 to 469 can be removed. I added these to remind myself how data is store. originally mizar does not have this.

jshaofuturewei replied:
Removed

pdgetrf commented on src/xdp/trn_transit_xdp.c:
```diff
@@ -454,6 +457,26 @@ static __inline int trn_process_inner_ip(struct transit_packet *pkt)
+	// send pkt to user space if this is the edge gateway and dest inner ip belongs to a virtual subnet
+	if ( net && net->virtual && pkt->ip->daddr == net->cluster_gateway) {
+		bpf_debug("[Transit:%d:0x%x] Edge gateway\n", __LINE__, bpf_ntohl(pkt->inner_ip->daddr));
+		// Rewrite RTS and update cache
```
this comment is probably copy pasted. the function of XDP_PASS here is to send pkt up to user space.

jshaofuturewei replied:
Updated

jshaofuturewei committed:
Addressed Peng's comments

pdgetrf commented on src/cli/test/test_cli.c:
```diff
@@ -1180,6 +1230,31 @@ static void test_trn_cli_update_net_subcmd(void **state)
 	expect_any(__wrap_update_net_1, clnt);
 	rc = trn_cli_update_net_subcmd(NULL, argc, argv1);
 	assert_int_equal(rc, -EINVAL);
+
```
side note: these test cases seem to have a lot of repetitions.

consider putting these in a loop.

vinaykul replied:
I agree, this is starting to get bloated. Please make this table driven. See example: https://github.com/CentaurusInfra/mizar/blob/dev-next/src/cli/test/test_cli.c#L1945

jshaofuturewei replied:
Updated

pdgetrf reviewed (approved):
LGTM

some more comments about a refactoring opportunity.

vinaykul commented on src/xdp/trn_transit_xdp.c:
```diff
@@ -454,6 +457,23 @@ static __inline int trn_process_inner_ip(struct transit_packet *pkt)
+	*/
+	struct network_key_t nkey;
+	struct network_t *net;
+	nkey.prefixlen = 96;
```
what's the hardcoded 96 for?

jshaofuturewei replied:
The hardcoded 96 is copied from https://github.com/CentaurusInfra/mizar/blob/dev-next/src/xdp/trn_transit_xdp.c#L144. The first 64 is the for the tunid and the rest 32 is for ip mask, I guess.

vinaykul commented on src/xdp/trn_transit_xdp.c:
```diff
@@ -454,6 +457,23 @@ static __inline int trn_process_inner_ip(struct transit_packet *pkt)
+	nkey.nip[2] = pkt->inner_ip->daddr;
+	net = bpf_map_lookup_elem(&networks_map, &nkey);
+	// send pkt to user space if this is the edge gateway and dest inner ip belongs to a virtual subnet
+	if ( net && net->virtual && pkt->ip->daddr == net->cluster_gateway) {
```
please remove whitespace between '(' 'net' for consistent coding style. prefer explicit net !=0 &&net->virtual != 0) && ... as these are not bool.

pdgetrf replied:
golang has lint (reformatting) and you can select to auto-format upon committing changes.

also net !=0 actually still can be confusing as to "why 0? what does 0 mean?". Probably add a #define for 0 with a meaningful name to make it clear what 0 means where.

jshaofuturewei replied:
bpf_map_lookup_elem  returns void * which is NULL if not found. I added the check here.

jshaofuturewei committed:
Addressed Vinay's comments

phudtran commented on out:
```diff
@@ -0,0 +1,4013 @@
+[2022-01-12 22:49:30,762] luigi-interface      [INFO    ] Loaded ['/etc/luigi/luigi.cfg']
```
Please remove this file

vinaykul replied:
why do we need this file?

vinaykul commented on src/cli/test/test_cli.c:
```diff
@@ -1107,79 +1107,177 @@ static void test_trn_cli_update_net_subcmd(void **state)
-	assert_int_equal(rc, -EINVAL);
+	struct test_case tc[] = {
+		{
+			"update_transit_pod_and_namespace_label_policy succeed with well formed endpoint json input",
```
we are not testing update_transit_pod_and_namespace_label_policy here

vinaykul reviewed (changes requested):
please see comments

jshaofuturewei committed:
Address vinay's comments

jshaofuturewei commented:
> please see comments

I addressed your comments. There are some issues with github that I can't reply my comments. I have to push my changes again. Please have a look again. Thanks

phudtran reviewed (approved):
Looks fine to me.

vinaykul commented on src/cli/test/test_cli.c:
```diff
@@ -1107,79 +1107,177 @@ static void test_trn_cli_update_net_subcmd(void **state)
 	uint32_t switches[] = { 0x100000a, 0x200000a };
-
+	/* Test cases */
 	struct rpc_trn_network_t exp_net = {
```
Where is this used?

vinaykul reviewed (commented):
One more comment. Please see if that struct is actually used.

jshaofuturewei commented:
> One more comment. Please see if that struct is actually used.

Removed

jshaofuturewei committed:
Removed a struct

vinaykul reviewed (approved):
looks good now
